File: src/main/java/inclub9/intellij/fieldlabel/FieldLabelAnnotator.java
package inclub9.intellij.fieldlabel;

import com.intellij.lang.annotation.AnnotationHolder;
import com.intellij.lang.annotation.Annotator;
import com.intellij.lang.annotation.HighlightSeverity;
import com.intellij.psi.*;
import org.jetbrains.annotations.NotNull;

public class FieldLabelAnnotator implements Annotator {
    @Override
    public void annotate(@NotNull PsiElement element, @NotNull AnnotationHolder holder) {
        if (!(element instanceof PsiAnnotation annotation)) {
            return;
        }

        if (!"FieldLabel".equals(annotation.getQualifiedName())) {
            return;
        }

        PsiAnnotationMemberValue value = annotation.findAttributeValue("value");
        if (value == null) {
            return;
        }

        // Validate that the annotation value matches either the field name or its uppercase version
        PsiField field = (PsiField) annotation.getParent();
        String fieldName = field.getName();
        String upperCaseName = camelCaseToUpperUnderscore(fieldName);

        String annotationValue = value.getText().replace("\"", "");
        if (!fieldName.equals(annotationValue) && !upperCaseName.equals(annotationValue)) {
            holder.newAnnotation(HighlightSeverity.WARNING,
                            "Field label value doesn't match field name or its uppercase version")
                    .create();
        }
    }

    private String camelCaseToUpperUnderscore(String input) {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < input.length(); i++) {
            char c = input.charAt(i);
            if (i > 0 && Character.isUpperCase(c)) {
                result.append('_');
            }
            result.append(Character.toUpperCase(c));
        }
        return result.toString();
    }
}

File: src/main/kotlin/inclub9/intellij/plugin/fieldlabel/annotator/FieldLabelAnnotator.kt
package inclub9.intellij.plugin.fieldlabel.annotator

import com.intellij.lang.annotation.AnnotationHolder
import com.intellij.lang.annotation.Annotator
import com.intellij.lang.annotation.HighlightSeverity
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiField
import com.intellij.psi.util.PsiUtil
import inclub9.intellij.plugin.fieldlabel.util.LombokStyle

class FieldLabelAnnotator : Annotator {
    override fun annotate(element: PsiElement, holder: AnnotationHolder) {
        if (element !is PsiField) return

        val annotation = element.annotations.find {
            it.qualifiedName == "inclub9.annotation.FieldLabel"
        } ?: return

        // Show the generated constant names in the editor
        val value = annotation.findAttributeValue("value")?.text?.trim('"')
        if (value != null) {
            val constantName = LombokStyle.camelCaseToUpperUnderscore(element.name)
            holder.newAnnotation(HighlightSeverity.INFORMATION,
                "Generates: $constantName = \"$value\"")
                .range(element.textRange)
                .create()
        }
    }
}
